docs/adr/2023-11-07_029-streaming-persistence.md

```diff
@@ -0,0 +1,40 @@
+---
+slug: 29
+title: |
+  29. EventServer abstraction for event persistence
+authors: [@cardenaso11, @quantumplation]
+tags: []
+---
+
+## Status
+Draft
+
+## Context
+
+The Hydra node represents a significant engineering asset, providing layer 1 monitoring, peer to peer consensus, durable persistence, and an isomorphic Cardano ledger. Because of this, it is being eyed as a key building block not just in Hydra based applications, but other protocols as well.
+
+One remaining difficulty in integrating Hydra into a fully productionalized software stack is the persistence model.  Currently, the Hydra node achieves durability by writing a sequence of "StateChanged" events to disk. If a node is interrupted, upon restarting, it can replay just these events, ignoring their corresponding effects, to recover the same internal state it had when it was interrupted. However, downstream consumers don't have the same luxury.
+
+We propose generalizing the persistence mechanism to open the door to a plugin based approach to extending the Hydra node.
+
+# Decision
+
+We propose adding a "persistence combinator", which can combine one or more "persistenceIncremental" instances.
+
+When appending to the combinator, it will forward the append to each persistence mechanism.
+
+As the node starts up, as part of recovering the node state, it will also ensure "at least once" semantics for each persistence mechanism. It will maintain a notion of a "primary" instance, from which events are loaded, and "secondary instances", which are given the opportunity to re-observe each event.
```

Comment:
If "secondary instances" are not even used to load anything.. are these even "persistence mechanisms"?

I do understand this idea rather as an "event sink".

Practically this would mean that instead of piggy-backing the `append` interface, we would want the node to submit events to zero or more event sinks.

I assume the _event_ in question is what is called `StateChanged` today + any information we currently model as `Effect` (at least `ClientEffects`). That is, any information which you could observe internally on the persisted `state` file + what is sent on the websocket today (`ServerOutput`).

Note that the term is heavily overloaded, but let's postulate with such a `data Event`.

A useful design IMO would be to have an

```hs
data EventSink m = EventSink { process :: [Event] -> m ()
```

handle, where the node has a list of those:

```hs
data HydraNode m = HydraNode { ..., eventSinks :: [EventSink m] }
```

and `stepHydraNode` (it's main loop) would process them in something like:

```hs
-- [...]
outcome <- logic
-- Not all hydra logic outcomes may yield and event
deriveEvents outcome $ \events ->
  forM_ eventSinks $ \EventSink{process} -> process events
```

That way, we could have multiple implementations for `EventSink`. I read sending things off to `S3` or through some IPC into a `scrolls` filter. 

We could even model the current API `data Server = Server { sendOutput :: ServerOutput -> m () }` as such an event sink.